### PR TITLE
[tc] Add count by bucket support on client side

### PR DESCRIPTION
**Pull request by LeoLeo718**
## Problem Statement
Currently, Venice thin client lacks the ability to perform bucket-based counting and aggregation operations on the client side. This limitation means that clients need to fetch all data and perform bucket aggregations manually, which is inefficient and can lead to unnecessary data transfer and processing overhead.

We need to implement a countGroupByBucket functionality that allows clients to:
- Count occurrences of values that match specific predicate conditions (buckets)
- Support all predicate types (IntPredicate, LongPredicate, FloatPredicate, DoublePredicate, StringPredicate)
- Support complex predicate combinations (AND, OR, nested combinations)
- Return bucket counts for multiple fields in a single request
- Handle all predicate methods (equalTo, greaterThan, greaterOrEquals, lowerThan, lowerOrEquals, anyOf)

## Solution
Implemented a new countGroupByBucket operation in the Venice thin client using pure client-side computation:

**Key Features:**
- **Predicate Support**: Handles all predicate types with comprehensive method coverage
- **Bucket Aggregation**: Groups data into buckets based on predicate conditions
- **Complex Combinations**: Supports AND/OR combinations and nested predicate logic
- **Null Handling**: Properly handles null values in predicate evaluation
- **Utf8 Conversion**: Automatically converts Avro Utf8 values to String for string predicates
- **Pure Client-Side**: Uses project() to fetch only required fields, then performs aggregation locally
- **Multi-Field Support**: Can aggregate multiple fields with different predicate types in a single request

**Implementation Details:**
- **AvroComputeAggregationRequestBuilder**: Builds aggregation requests with validation:
  - Validates bucket definitions (non-null, non-empty bucket maps)
  - Validates field names (non-null, non-empty, must exist in schema)
  - Projects required fields for client-side processing
  - Supports chaining multiple countGroupByBucket calls
- **AvroComputeAggregationResponse**: Processes results:
  - Evaluates predicates against field values
  - Aggregates counts across all compute results for each bucket
  - Handles Utf8 to String conversion for string fields
  - Returns bucket name to count mappings
  - Supports generic type casting with proper error handling

### Code changes
- [x] Added new code behind **existing interfaces**. No new configs needed.
- [ ] Introduced new **log lines**. 
  - [ ] Confirmed if logs need to be **rate limited** to avoid excessive logging.

### **Concurrency-Specific Checks**
Both reviewer and PR author to verify
- [x] Code has **no race conditions** or **thread safety issues**.
- [x] Proper **synchronization mechanisms** (e.g., `synchronized`, `RWLock`) are used where needed.
- [x] No **blocking calls** inside critical sections that could lead to deadlocks or performance degradation.
- [x] Verified **thread-safe collections** are used (e.g., `ConcurrentHashMap`, `CopyOnWriteArrayList`).
- [x] Validated proper exception handling in multi-threaded code to avoid silent thread termination.

**Concurrency Details:**
- Each request creates its own response object - no shared state
- Result aggregation happens in request-scoped maps
- Predicate evaluation is stateless and thread-safe
- No blocking calls or synchronization needed - pure computation
- Thread-safe by design - immutable request builders and response objects

## How was this PR tested?
- [x] New unit tests added.
- [x] New integration tests added.
- [x] Modified or extended existing tests.
- [x] Verified backward compatibility (if applicable).

**Unit tests added:**
- **AvroComputeAggregationRequestBuilderTest**: Comprehensive validation tests
  - Valid configurations with single and multiple fields
  - Invalid bucket definitions (null, empty)
  - Null and empty field validation
  - Non-existent field detection
  - Chaining multiple countGroupByBucket calls
  - Boundary cases for field names and bucket names

- **AvroComputeAggregationResponseTest**: Result processing tests
  - All predicate type methods (IntPredicate, LongPredicate, FloatPredicate, DoublePredicate, StringPredicate)
  - Complex predicate combinations (AND, OR, nested)
  - Null value handling in predicate evaluation
  - Utf8 value conversion for string fields
  - Type casting error scenarios
  - Mixed data types handling
  - Edge cases for all predicate methods

**Integration tests added:**
- **ReadComputeValidationTest.testCountGroupByBucketAggregation**: End-to-end test
  - Creates store with multiple field types (int, long, float, double, string)
  - Populates with test data covering various scenarios
  - Verifies bucket aggregation on all field types
  - Tests all predicate methods for each type
  - Verifies complex predicate combinations
  - Tests with various data patterns and bucket definitions

**Test Coverage:**
- Input validation: 100% coverage of error cases
- Response processing: All predicate types and methods covered
- Integration: Full workflow from request to response validated
- Type casting: Proper error handling for type mismatches
- Predicate evaluation: All combination scenarios tested
- Edge cases: Null values, empty results, boundary conditions

## Does this PR introduce any user-facing or breaking changes?
- [x] No. You can skip the rest of this section.
- [ ] Yes. Clearly explain the behavior change and its impact.

This is a new feature that doesn't affect existing functionality:
- All changes are backward compatible
- New functionality is opt-in through the new countGroupByBucket API
- Existing client operations remain unchanged
- No changes to wire protocol or storage format
- Existing countGroupByValue functionality remains unaffected

### CODE REVIEW

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      if (entry.getKey() == null || entry.getKey().isEmpty()) {
+        throw new VeniceClientException("Bucket name cannot be null or empty");
+      }
+      if (entry.getValue() == null) {
+        throw new VeniceClientException("Predicate for bucket '" + entry.getKey() + "' cannot be null");
+      }
+    }
+
+    // Validate fields exist in schema
+    Schema valueSchema = schemaReader.getValueSchema(schemaReader.getLatestValueSchemaId());
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java:89]: Maybe we could define a private function for this validation logic, it is shared with countGroupByValue

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        Predicate predicate = bucketEntry.getValue();
+
+        try {
+          boolean matches = false;
+
+          // Convert field value if needed
+          Object convertedValue = convertUtf8ToString(fieldValue);
+
+          // Handle different value types
+          if (convertedValue instanceof GenericRecord) {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: Here the if checking logic is a bit complicated. As we define predicate for each type for a specific type, so I think it is important to let user follow the strong type indication for predicate. We could also add validation logic in request builder to check schema to ensure predicator type matching with field value type to prevent type mismatch happen. What do you think?

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    for (String fieldName: fieldNames) {
+      if (fieldName == null) {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Maybe we could define a private function for this validation logic, it is shared with countGroupByValue

**Review by haoxu07 — COMMENTED** [review]
Thanks for preparing this! Just a quick pass, will check tests later.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java (hunk 2 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: I think it will be better since that will ensure if a user uses the wrong type, the problem will be detected immediately during the construction request, allowing for early failure and early correction.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 3 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Oh! You're right!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 4]
```diff
@@ -26,29 +31,121 @@ public AvroComputeAggregationRequestBuilder(
...
+      // For generic predicates, we allow them to work with any type
+      // The actual type checking will happen at runtime
+      return true;
+    }
+  }
+
+  /**
+   * Gets a human-readable description of the predicate type.
+   */
+  private String getPredicateType(Predicate<?> predicate) {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: This is mainly for logging purpose? If so, we could explore by adding toString functionality to Predicate implementations.

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java [hunk 5]
```diff
@@ -240,6 +241,430 @@ public void testTypeCastingThrowsException() {
     System.identityHashCode((Integer) firstKey);
   }
 
+  // ========== countByBucket Tests ==========
+
+  @Test(description = "Should handle all IntPredicate methods correctly")
+  public void testIntPredicateMethods() {
+    Map<String, ComputeGenericRecord> computeResults = new HashMap<>();
+    Map<String, Map<String, com.linkedin.venice.client.store.predicate.Predicate>> fieldBucketMap = new HashMap<>();
```
Comment (haoxu07) [inline clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java]: You should be able to import the class: com.linkedin.venice.client.store.predicate.Predicate, and same as below.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 6]
```diff
@@ -26,29 +31,121 @@ public AvroComputeAggregationRequestBuilder(
...
+    }
+  }
+
+  /**
+   * Checks if the predicate type is compatible with the given Avro schema type.
+   */
+  private boolean isPredicateTypeCompatible(Predicate<?> predicate, Schema schema) {
+    Schema.Type avroType = schema.getType();
+
+    if (predicate instanceof LongPredicate) {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Seems predicate type enforcement is less strong. For example, we want `DoublePredicate` to support float, int and double. Not sure if this is ideal. IMO, it is better that following strong type eforcement is OK, as value schema is defined by user and predicate type is also provided by user. It is users' responsibility to get them align and we could fail if it does not meet.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 7]
```diff
@@ -26,29 +31,121 @@ public AvroComputeAggregationRequestBuilder(
...
+                  getPredicateType(predicate)));
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if the predicate type is compatible with the given Avro schema type.
+   */
+  private boolean isPredicateTypeCompatible(Predicate<?> predicate, Schema schema) {
```
Comment (FelixGV) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java:98]: This should be a function on the predicate, and leverage polymorphism, rather than a series of if instanceof.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 4 above)
Comment (FelixGV) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: `predicate.getClass().getSimpleName()` would be equivalent... no need for a function.

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Will move to predicate!

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java (hunk 5 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java]: ok

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 6 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Yeah, you're right!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 7 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java:98]: great! I will move it to predicate

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 4 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Abosolutely.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 8]
```diff
@@ -53,6 +63,139 @@ public <T> Map<T, Integer> getValueToCount(String field) {
...
+          Object valueToEvaluate = convertedValue;
+          if (predicate instanceof LongPredicate) {
+            valueToEvaluate = convertToType(convertedValue, Long.class);
+          } else if (predicate instanceof IntPredicate) {
+            valueToEvaluate = convertToType(convertedValue, Integer.class);
+          } else if (predicate instanceof FloatPredicate) {
+            valueToEvaluate = convertToType(convertedValue, Float.class);
+          } else if (predicate instanceof DoublePredicate) {
+            valueToEvaluate = convertToType(convertedValue, Double.class);
+          }
```
Comment (FelixGV) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:110]: What is this for? Like... if a user specifies a `LongPredicate` on a `String` field, then we'll try to parse the content of the string to convert it into a long? If that is the goal, the IDK if it's a good idea... on one hand, it's very permissive, which seems like it could be convenient, but on the other hand, it must be best-effort, since a string is not guaranteed to be convertible to a long, and therefore it could be brittle... I see that you catch exceptions and return null in `convertToType`, which may be better than blowing up, in some sense, but it could lead to confusing behavior.

I don't mind leaving it like this for now, but I think we should consider making it more strict, e.g., if a predicate is defined on a field with an incompatible type, then we should fail fast at query definition-time, rather than failing (or swallowing...) later on during query evaluation-time. We could also consider offering various modes, e.g., `enum TypeCompatibility { STRICT, PERMISSIVE }`. This here would be the permissive one, and there could be another one which is strict.

One more thing, similar to other comments I made... let's try as much as possible to leverage polymorphism. `convertToType` could be a function of the predicate, this way we eliminate branching both here and in the `convertToType` implementation below. (The branching is still there at run-time, but it's hidden by the type system, and becomes an implementation detail of how the JVM picks the right class to execute, rather than being laid out explicitly in the code).

Anyway... like I said, we can leave it like this for now, but let's just keep the above things in mind as potential improvements for the future.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 9]
```diff
@@ -8,17 +13,26 @@
 
 
 /**
- * Implementation of {@link ComputeAggregationResponse} that handles the results of count-by-value aggregations.
+ * Implementation of {@link ComputeAggregationResponse} that handles the results of count-by-value and count-by-bucket aggregations.
  */
 public class AvroComputeAggregationResponse<K> implements ComputeAggregationResponse {
   private final Map<K, ComputeGenericRecord> computeResults;
   private final Map<String, Integer> fieldTopKMap;
```
Comment (FelixGV) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:20]: So this is for the `countByValue` operations, right? We might want to mention that in the field name, to be more explicit.
